vivisect/analysis/aarch64/emulation.py

```diff
@@ -0,0 +1,486 @@
+import logging
+
+import visgraph.pathcore as vg_path
+
+import vivisect
+import vivisect.exc as v_exc
+import vivisect.impemu.monitor as viv_monitor
+import vivisect.analysis.generic.codeblocks as viv_cb
+
+import envi
+import envi.archs.aarch64 as e_a64
+
+from envi.archs.aarch64.regs import *
+from envi.archs.aarch64.const import *
+
+from vivisect.const import *
+
+logger = logging.getLogger(__name__)
+
+
+class AnalysisMonitor(viv_monitor.AnalysisMonitor):
+
+    def __init__(self, vw, fva, verbose=True):
+        viv_monitor.AnalysisMonitor.__init__(self, vw, fva)
+        self.retbytes = None
+        self.badops = vw.arch.archGetBadOps()
+        self.last_lr_pc = 0
+        self.strictops = False
+        self.returns = False
+        self.infloops = []
+        self.switchcases = 0
+
+    def prehook(self, emu, op, starteip):
+
+        try:
+            tmode = emu.getFlag(PSR_T_bit)
+            self.last_tmode = tmode
+            if op in self.badops:
+                emu.stopEmu()
+                raise v_exc.BadOpBytes(op.va)
+
+            viv_monitor.AnalysisMonitor.prehook(self, emu, op, starteip)
+
+            loctup = emu.vw.getLocation(starteip)
+            if loctup is None:
+                # logger.debug("emulation: prehook: new LOC_OP  fva: 0x%x     starteip: 0x%x  flags: 0x%x", self.fva, starteip, op.iflags)
+                arch = (envi.ARCH_ARMV7, envi.ARCH_THUMB)[(starteip & 1) | tmode]
+                emu.vw.makeCode(starteip & -2, arch=arch)
```

Comment:
I know you do something similar in the base ARM, but should this be more of an extension of the current function (should we be in one)?